### ISSUE TITLE
[MPS] [1.13.0 regression] autograd returns NaN loss, originating from NativeGroupNormBackward0

### ISSUE THREAD

**Issue by Birch-san** [pytorch/pytorch#88331]
### 🐛 Describe the bug

`x + GroupNorm()(x)` stacked enough times seems to result in NaN gradients' being returned by autograd.

affects stable-diffusion. breaks CLIP guidance. I believe this explains also why textual inversion training encounters immediate [NaN loss on 1.13.0, despite working fine on 1.12.1](https://github.com/invoke-ai/InvokeAI/issues/517#issuecomment-1300447429). this may explain the NaNs I encountered on nightlies [as long ago as Sep 22](https://github.com/invoke-ai/InvokeAI/issues/517#issuecomment-1255630588). if it impacts textual inversion training: it probably impacts stable-diffusion training more generally (for example anything which backpropagates through the VAE's decoder).

CLIP guidance comparison…  
PyTorch 1.12.1:  
<img width="64" src="https://user-images.githubusercontent.com/6141784/199550570-d45135d2-b5dc-4392-b63a-24b1bfab6588.png"> ➔ <img width="64" src="https://user-images.githubusercontent.com/6141784/199547727-87f3fcac-5a6d-4b1d-936e-6898bccc9c69.png">  

PyTorch 1.13.0 (latents are NaN):  
<img width="64" src="https://user-images.githubusercontent.com/6141784/199550570-d45135d2-b5dc-4392-b63a-24b1bfab6588.png"> ➔ <img width="64" src="https://user-images.githubusercontent.com/6141784/199547826-607d87ec-6728-4532-a986-7ec41799a48b.png">

chaining `x + GroupNorm()(x)`s is a pattern which occurs in ResnetBlocks.  

```python
import torch

from torch import __version__, Tensor, enable_grad, autograd
from torch.nn import functional as F, Conv2d, GroupNorm

device = torch.device('mps')

print(f'torch.__version__: {__version__}')
shape = [1, 4, 16, 16]
x = torch.full(shape, 7.0, device=device)

target = torch.ones((1, 3, 128, 128), device=device)

conv_in = Conv2d(4, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), device=device)
conv_out = Conv2d(128, 3, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), device=device)
norm = GroupNorm(32, 128, eps=1e-6, affine=True, device=device)

with enable_grad():#, autograd.detect_anomaly():
    x = x.detach().requires_grad_()

    out = 5.5 * x

    out = conv_in(out)
    out = out+norm(out)
    out = out+norm(out)
    out = out+norm(out)
    out = F.interpolate(out, scale_factor=8.0, mode="nearest")
    out = norm(out)
    out = conv_out(out)

    loss: Tensor = (out - target).norm(dim=-1).sum()
    grad: Tensor = -autograd.grad(loss, x)[0]
    assert not grad.detach().isnan().any().item(), 'NaN gradients returned by autograd'

print('Ran to completion successfully; did not encounter NaN gradients.')
```

problem reproduces on MPS on 1.13.0.  
does not reproduce on CPU on 1.13.0.  
does not reproduce on MPS on 1.12.1.

<img width="1407" alt="image" src="https://user-images.githubusercontent.com/6141784/199552931-144fc038-3fa1-40b8-9d9c-2917874b48c2.png">

### Versions

```
PyTorch version: 1.13.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.5 (arm64)
GCC version: Could not collect
Clang version: 13.0.0 (clang-1300.0.29.30)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.10.6 (main, Oct 24 2022, 11:04:07) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-12.5-arm64-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] clip-anytorch==2.5.0
[pip3] numpy==1.23.3
[pip3] open-clip-torch==2.0.2
[pip3] pytorch-lightning==1.7.7
[pip3] torch==1.13.0
[pip3] torchdiffeq==0.2.3
[pip3] torchmetrics==0.10.1
[pip3] torchvision==0.14.0
[conda] clip-anytorch             2.5.0                    pypi_0    pypi
[conda] numpy                     1.23.3          py310h220015d_0  
[conda] numpy-base                1.23.3          py310h742c864_0  
[conda] open-clip-torch           2.0.2                    pypi_0    pypi
[conda] pytorch                   1.13.0                 py3.10_0    pytorch
[conda] pytorch-lightning         1.7.7                    pypi_0    pypi
[conda] torchdiffeq               0.2.3                    pypi_0    pypi
[conda] torchmetrics              0.10.1                   pypi_0    pypi
[conda] torchvision               0.14.0                py310_cpu    pytorch
```

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by albanD** [comment]
High priority to find the root cause.

**Comment by malfet** [comment]
Just curious: does it affect trunk?

**Comment by Birch-san** [comment]
yes, happens on `1.14.0.dev20221102` too.

**Comment by Birch-san** [comment]
happens on `1.13.0.dev20220904` too. that's the oldest nightly I could find on https://download.pytorch.org/whl/nightly/torch/.

**Comment by kulinseth** [comment]
@Birch-san , I think there were changes in the Core where GroupNorm layer support was moved to   `aten::native_group_norm_backward` op. We don't have support for that in MPS. If you are running with FALLBACK enabled, there could be bug in that path itself.

**Comment by Birch-san** [comment]
yeah, entirely possible.

if I run on 1.12.1: the reproducer passes without needing to specify `PYTORCH_ENABLE_MPS_FALLBACK=1`.

but on 1.13.0: the reproducer only runs with `PYTORCH_ENABLE_MPS_FALLBACK=1`, and warns:

> /Users/birch/anaconda3/envs/sd/lib/python3.10/site-packages/torch/autograd/__init__.py:300: UserWarning: The operator 'aten::native_group_norm_backward' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)

so yes, it's possible it could be related to the fallback-to-CPU path, or the CPU implementation of `aten::native_group_norm_backward`.  
so it's not that that _op_ has regressed, but rather that the unsupported op was not in the path of this algorithm before — and now is — and consequently the use-case has regressed.

**Comment by kulinseth** [comment]
> yeah, entirely possible.
> 
> if I run on 1.12.1: the reproducer passes without needing to specify `PYTORCH_ENABLE_MPS_FALLBACK=1`.
> 
> but on 1.13.0: the reproducer only runs with `PYTORCH_ENABLE_MPS_FALLBACK=1`, and warns:
> 
> > /Users/birch/anaconda3/envs/sd/lib/python3.10/site-packages/torch/autograd/**init**.py:300: UserWarning: The operator 'aten::native_group_norm_backward' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
> 
> so yes, it's possible it could be related to the fallback-to-CPU path, or the CPU implementation of `aten::native_group_norm_backward`. so it's not that that _op_ has regressed, but rather that the unsupported op was not in the path of this algorithm before — and now is — and consequently the use-case has regressed.

Right, we are planning to add the op as it affects some other torchbench networks and requested in the op coverage GitHub issue . Will update here with pr to see if that helps .

**Comment by malfet** [comment]
IMO there are 3 separate issues in play:
 - We regression was not noticed in time (i.e. there should be a test validating `GroupNorm` forward and backward on MPS)
 - Why fallback to MPS returns NaNs
 - Implementing backward pass natively (I wonder if we could use decomps for this one)

**Comment by atalman** [comment]
tentative for release. If fix is not provided we will drop it from the minor release 1.13.1

**Comment by atalman** [comment]
@kulinseth We are planning on including this fix in the release 1.13.1.
Could you please provide ETA for this issue ?

Following are the tentative dates for 1.13.1:
Phase 1 (until 12/8/2022): Release Candidate 1
Phase 2 (until 12/15/2022): Tentative General Availability

**Comment by atalman** [comment]
Removed from 1.13.1 since the fix was not provided by 12.8.2022

**Comment by kulinseth** [comment]
> Removed from milestone 1.13.1 since the fix was not provided by 12.8.2022

Hi @atalman , yes makes sense. It will be fixed in nightly. cc @malfet , are you planning to take a look at the decomps route to fix it ? Alternatively we can register the NativeGroupNorm op.

**Comment by malfet** [comment]
@kulinseth is there a native MPS implementation? Or would it be a composite op anyway?
The Python code I plan to use are implemented here:
https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/torch/_decomp/decompositions.py#L1146-L1158

**Comment by kulinseth** [comment]
> @kulinseth is there a native MPS implementation? Or would it be a composite op anyway? The Python code I plan to use are implemented here:
> 
> https://github.com/pytorch/pytorch/blob/eeb3f8aa5496f713339fea678d145575c884e459/torch/_decomp/decompositions.py#L1146-L1158

@malfet , it will be composite op anyway. Using the decomposition path sounds reasonable to me.

**Comment by malfet** [comment]
Ok, some more info: problem is not the backward pass, but `grad_output` argument, which is passed as input

**Comment by malfet** [comment]
Ok, so following registers enough prims to make this code works correctly in trunk:
```
_lib = torch.library.Library("aten", "IMPL")
_lib.impl("native_group_norm_backward", torch._decomp.decompositions.native_group_norm_backward, "MPS")
_lib.impl("native_group_norm", torch._refs.native_group_norm, "MPS")
_lib.impl("var_mean.correction", torch._refs.var_mean, "MPS")
```
